### DISCUSSION TITLE
fixture teardown sequence

### DISCUSSION THREAD

**Discussion by ws1088** [pytest-dev/pytest#9100] [Q&A]
Hi all,
It looks like a function scope fixture is torn down after the printing of `.` in the console. I am writing a fixture which capture stdout/stderr to a file and it is capturing `�[32m.�[0m` before `print('hello')`:

```
@pytest.fixture
def my_fixture(request: FixtureRequest):
    # setup capturing of stdout/stderr to a test-specific file
    # ...
    yield 'something'
    print('hello')
    # restore stdout/stderr
    # ...
```

test-specific file content:
```
[32m.[0mhello
```

Am I doing anything wrong?
Thanks in advance.

**Comment by portyu9** [comment]
Hi @ws1088 

What you are seeing is consistent with pytest's execution order.

With a yield fixture:

```python
setup
yield
teardown
```

pytest runs the test call after the `yield`, generates/reports the test-call result, and only then resumes fixture teardown.

So if your fixture redirects the process's stdout/file descriptor **across the entire `yield`**, pytest's own progress output such as:

```text
.
```

can be written while that redirection is still active. By the time the fixture restores stdout in teardown, the progress character has already gone to your file.

The fix is not to hold a global stdout/stderr redirection open across the entire pytest test lifecycle.

Prefer one of:

- pytest's `capsys` / `capfd` fixtures for captured output;
- an application-specific output stream;
- a logging `FileHandler`;
- a narrowly scoped redirect around only the application call that needs it.

A fixture finalizer cannot restore the descriptor before pytest reports the test call because teardown happens afterward.

Please mark this answer as accepted if it helped, thank you.
